extensions/microsoft-authentication/src/node/authProvider.ts

```diff
@@ -187,86 +186,70 @@ export class MsalAuthProvider implements AuthenticationProvider {
+				return;
 			}
-			// This error comes from the backend and is likely not due to the user's machine
-			// failing to open a port or something local that would require us to try the
-			// URL handler loopback client.
-			if (e instanceof ServerError) {
-				this._telemetryReporter.sendLoginFailedEvent();
-				throw e;
+			const message = userCancelled
+				? l10n.t('Having trouble logging in? Would you like to try a different way? ({0})', mode)
+				: l10n.t('You have not yet finished authorizing this extension to use your Microsoft Account. Would you like to try a different way? ({0})', mode);
+			const result = await window.showWarningMessage(message, yes, no);
+			if (result !== yes) {
+				throw new CancellationError();
 			}
+		};
 
-			// The user closed the modal window
-			if ((e as ClientAuthError).errorCode === ClientAuthErrorCodes.userCanceled) {
-				this._telemetryReporter.sendLoginFailedEvent();
-				throw e;
-			}
+		const flows = getMsalFlows({
+			extensionHost: typeof navigator === 'undefined'
+				? this._context.extension.extensionKind === ExtensionKind.UI ? ExtensionHost.Local : ExtensionHost.Remote
+				: ExtensionHost.WebWorker,
+		});
 
-			// The user wants to try the loopback client or we got an error likely due to spinning up the server
-			const loopbackClient = new UriHandlerLoopbackClient(this._uriHandler, redirectUri, this._logger);
+		let lastError: Error | undefined;
+		for (const flow of flows) {
+			if (flow !== flows[0]) {
+				try {
+					await promptToContinue(flow.label);
+				} finally {
+					this._telemetryReporter.sendLoginFailedEvent();
+				}
+			}
 			try {
-				const windowHandle = window.nativeHandle ? Buffer.from(window.nativeHandle) : undefined;
-				result = await cachedPca.acquireTokenInteractive({
-					openBrowser: (url: string) => loopbackClient.openBrowser(url),
+				const result = await flow.trigger({
+					cachedPca,
 					scopes: scopeData.scopesToSend,
-					loopbackClient,
 					loginHint: options.account?.label,
-					prompt: options.account?.label ? undefined : 'select_account',
-					windowHandle
+					windowHandle: window.nativeHandle ? Buffer.from(window.nativeHandle) : undefined,
+					logger: this._logger,
+					uriHandler: this._uriHandler
 				});
+
+				const session = this.sessionFromAuthenticationResult(result, scopeData.originalScopes);
+				this._telemetryReporter.sendLoginEvent(session.scopes);
+				this._logger.info('[createSession]', `[${scopeData.scopeStr}]`, 'returned session');
+				this._onDidChangeSessionsEmitter.fire({ added: [session], changed: [], removed: [] });
+				return session;
 			} catch (e) {
-				this._telemetryReporter.sendLoginFailedEvent();
-				throw e;
+				lastError = e;
+				if (e instanceof ServerError || (e as ClientAuthError)?.errorCode === ClientAuthErrorCodes.userCanceled) {
+					this._telemetryReporter.sendLoginFailedEvent();
+					throw e;
+				}
+				// Continue to next flow
+				if (e instanceof CancellationError) {
```

Comment:
Nit: could be

```suggestion
			userCancelled = (e instanceof CancellationError);
```

Also do we need to re-throw in the cancellation case too?